File: MockBulkLoadFileImporter.vb
Imports System.Threading
Imports kCura.EDDS.WebAPI.BulkImportManagerBase
Imports kCura.WinEDDS.TApi
Imports Relativity.Logging

Namespace kCura.WinEDDS.NUnit
	''' <summary>
	''' Inheriting the BulkLoadFileImporter to override its BulkImportManager property (the part we want to throw timeout exceptions)
	''' </summary>
	Public Class MockBulkLoadFileImporter
		Inherits kCura.WinEDDS.BulkLoadFileImporter

		Private Property WillThrowException As Boolean

		Public Property FieldMap As LoadFileFieldMap
			Get
				Return MyBase._fieldMap
			End Get
			Set(ByVal value As LoadFileFieldMap)
				MyBase._fieldMap = value
			End Set
		End Property

		Public Property BatchSize As Int32
			Get
				Return Me.ImportBatchSize
			End Get
			Set(ByVal value As Int32)
				Me.ImportBatchSize = value
			End Set
		End Property

		Protected Overrides ReadOnly Property BatchResizeEnabled As Boolean
			Get
				Return True
			End Get
		End Property

		Public PauseCalled As Int32 = 0

		Public Sub New(ByVal args As LoadFile, ByVal processController As kCura.Windows.Process.Controller, ByRef ioReporterInstance As IIoReporter, ByRef logger As ILog, 
					   ByVal timeZoneOffset As Int32, ByVal autoDetect As Boolean, ByVal initializeUploaders As Boolean, ByVal processID As Guid, ByVal doRetryLogic As Boolean, ByVal bulkLoadFileFieldDelimiter As String, ByVal throwsException As Boolean, ByVal bulkManager As kCura.WinEDDS.Service.BulkImportManager,  ByRef tokenSource As CancellationTokenSource)
			MyBase.new(args, processController, ioReporterInstance, logger, timeZoneOffset, autoDetect, initializeUploaders, processID, doRetryLogic, bulkLoadFileFieldDelimiter, False, tokenSource)
			Me.WillThrowException = throwsException
			_bulkImportManager = bulkManager
			Me.ImportBatchSize = 500
			Me.ImportBatchVolume = 1000000
		End Sub

		Public Function CleanFolderPath(ByVal path As String) As String
			Return MyBase.CleanDestinationFolderPath(path)
		End Function

		Public Function ConvertOverlayBehaviorEnum(ByVal inputOverlayType As LoadFile.FieldOverlayBehavior?) As kCura.EDDS.WebAPI.BulkImportManagerBase.OverlayBehavior
			Return MyBase.GetMassImportOverlayBehavior(inputOverlayType)
		End Function

		Public Function TryBulkImport(ByVal settings As NativeLoadInfo) As MassImportResults
			Return BulkImport(settings, True)
		End Function

		Protected Overrides ReadOnly Property BulkImportManager As kCura.WinEDDS.Service.BulkImportManager
			Get
				If _bulkImportManager Is Nothing Then _bulkImportManager = New MockBulkImportManagerWebExceptions(Me.WillThrowException)
				Return _bulkImportManager
			End Get
		End Property

		Protected Overrides Sub RaiseWarningAndPause(ByVal ex As System.Exception, ByVal timeoutSeconds As Integer)
			PauseCalled += 1
		End Sub

		Public Property MinimumBatch As Integer
			Get
				Return MyBase.MinimumBatchSize
			End Get
			Set(ByVal value As Integer)
				MinimumBatchSize = value
			End Set
		End Property

		Protected Overrides ReadOnly Property ParentArtifactTypeID As Integer
			Get
				Return 4000
			End Get
		End Property

		Protected Overrides ReadOnly Property NumberOfRetries As Int32
			Get
				Return 3
			End Get
		End Property

	End Class
End NameSpace

File: Source/Relativity.DataExchange.Import.NUnit/MockBulkLoadFileImporter.cs
// -----------------------------------------------------------------------------------------------------
// <copyright file="MockBulkLoadFileImporter.cs" company="Relativity ODA LLC">
//   © Relativity All Rights Reserved.
// </copyright>
// -----------------------------------------------------------------------------------------------------

namespace Relativity.DataExchange.Import.NUnit
{
	using System;
	using System.Net;
	using System.Threading;
	using kCura.EDDS.WebAPI.BulkImportManagerBase;
	using kCura.WinEDDS;

	using Moq;

	using Relativity.DataExchange.Io;
	using Relativity.DataExchange.Process;
	using Relativity.DataExchange.Service;
	using Relativity.DataExchange.TestFramework;
	using Relativity.DataExchange.Transfer;
	using Relativity.Logging;

	/// <summary>
	/// Represents a mock class object for <see cref="BulkLoadFileImporter"/>.
	/// </summary>
	public sealed class MockBulkLoadFileImporter : BulkLoadFileImporter
	{
		public MockBulkLoadFileImporter(
			LoadFile args,
			ProcessContext context,
			IIoReporter reporter,
			ILog logger,
			int timeZoneOffset,
			bool autoDetect,
			bool initializeUploaders,
			Guid processID,
			bool doRetryLogic,
			string bulkLoadFileFieldDelimiter,
			kCura.WinEDDS.Service.Replacement.IBulkImportManager manager,
			CancellationTokenSource tokenSource,
			IRunningContext runningContext)
			: base(
				args,
				context,
				reporter,
				logger,
				timeZoneOffset,
				autoDetect,
				initializeUploaders,
				processID,
				doRetryLogic,
				bulkLoadFileFieldDelimiter,
				tokenSource,
				() => nameof(MockBulkLoadFileImporter),
				runningContext)
		{
			this._bulkImportManager = manager;
			this.ImportBatchSize = 500;
			this.ImportBatchVolume = 1000000;
			this.OutputFileWriter = new OutputFileWriter(
				new TestNullLogger(),
				Relativity.DataExchange.Io.FileSystem.Instance);
			this.OutputFileWriter.Open(true);
		}

		public int GetMetadataFilesCount => this.MetadataFilesCount;

		public int BatchSize
		{
			get => this.ImportBatchSize;
			set => this.ImportBatchSize = value;
		}

		public LoadFileFieldMap FieldMap
		{
			get => this._fieldMap;
			set => this._fieldMap = value;
		}

		public int MinimumBatch
		{
			get => this.MinimumBatchSize;
			set => this.MinimumBatchSize = value;
		}

		public int PauseCalled
		{
			get;
			private set;
		}

		protected override bool BatchResizeEnabled => true;

		protected override int ParentArtifactTypeID => 4000;

		protected override int NumberOfRetries => 3;

		protected override int WaitTimeBetweenRetryAttempts => 0;

		[System.Diagnostics.CodeAnalysis.SuppressMessage(
			"Microsoft.Performance",
			"CA1822:MarkMembersAsStatic",
			Justification = "To be consistent with existing VB.NET tests.")]
		public string CleanFolderPath(string path)
		{
			return CleanDestinationFolderPath(path);
		}

		public OverlayBehavior ConvertOverlayBehavior(LoadFile.FieldOverlayBehavior? behavior)
		{
			return this.GetMassImportOverlayBehavior(behavior);
		}

		public MassImportResults TryBulkImport(NativeLoadInfo settings)
		{
			const bool IncludeExtractedTextEncoding = true;
			return this.BulkImport(settings, IncludeExtractedTextEncoding);
		}

		public void SetTapiBridges()
		{
			UploadTapiBridgeParameters2 parameters = new UploadTapiBridgeParameters2
				                                         {
					                                         Credentials = new NetworkCredential(),
					                                         WebServiceUrl = "https://relativity.one.com",
					                                         WorkspaceId = 1337,
					                                         TargetPath = "./",
					                                         FileShare = "./somepath/",
					                                         TimeoutSeconds = 0,
				                                         };
			var relativityManagerServiceFactoryMock = new Mock<IRelativityManagerServiceFactory>();
			relativityManagerServiceFactoryMock
				.Setup(x => x.Create(It.IsAny<RelativityInstanceInfo>(), It.IsAny<bool>()))
				.Returns<RelativityInstanceInfo, bool>((x, y) => new RelativityManagerService(x));
			this.CreateTapiBridges(parameters, parameters.ShallowCopy(), new NullAuthTokenProvider(), relativityManagerServiceFactoryMock.Object);
		}

		public void SetBatchCounter(int numberToSet)
		{
			this._batchCounter = numberToSet;
		}

		public void PushNativeBatchInvoker(string outputNativePath, bool shouldCompleteJob, bool lastRun)
		{
			this.PushNativeBatch(outputNativePath, shouldCompleteJob, lastRun);
		}

		public override void Dispose()
		{
			this.OutputFileWriter?.Dispose();
			base.Dispose();
		}

		protected override void RaiseWarningAndPause(Exception exception, int timeoutSeconds, int retryCount, int totalRetryCount)
		{
			base.RaiseWarningAndPause(exception, timeoutSeconds, retryCount, totalRetryCount);
			this.PauseCalled++;
		}
	}
}